
<h1 style="color: #9f43c3ff; border-bottom: 3px solid #9f43c3ff; padding-bottom: 8px;">

AI & ML Course with BinX Week 3 Day 1 : Introduction to Scikit-Learn & Data Splitting

</h1>

<blockquote style="border-left: 3px solid #2e1a9aff; padding-left: 12px; margin-left: 0;">

<b>Day 1 Learning Objectives:</b>
- 📂 Load and explore tabular data using Pandas.
- 🎯 Define the Feature Matrix ($X$) and Target Vector ($y$).
- ✂️ Implement <code>train_test_split</code> to partition dataset into training (80%) and testing (20%) sets.
- 🔒 Ensure experiment reproducibility using <code>random_state=42</code>.

</blockquote>

## <span style="color: #39b64eff">1.1 Environment Setup & Library Imports</span>

<blockquote style="border-left: 3px solid #FD1D1D; padding-left: 12px; margin-left: 0;">

We begin by importing core numerical, data manipulation, and visualization libraries alongside Scikit-Learn's <code>train_test_split</code> utility.

</blockquote>

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split

print("Libraries successfully imported.")

Libraries successfully imported


## <span style="color: #309c42ff">1.2 Dataset Loading & Exploration</span>

<blockquote style="border-left: 3px solid #FD1D1D; padding-left: 12px; margin-left: 0;">

We load the Palmer Penguins dataset from <code>../../Data/raw/penguins_size.csv</code> and inspect the initial records, dataset shape, and feature columns.

</blockquote>

In [ ]:

df = pd.read_csv("../../Data/raw/penguins_size.csv")

df.head()

,species,island,culmen_length_mm,culmen_depth_mm,flipper_length_mm,body_mass_g,sex
0,Adelie,Torgersen,39.1,18.7,181.0,3750.0,MALE
1,Adelie,Torgersen,39.5,17.4,186.0,3800.0,FEMALE
2,Adelie,Torgersen,40.3,18.0,195.0,3250.0,FEMALE
3,Adelie,Torgersen,NaN,NaN,NaN,NaN,NaN
4,Adelie,Torgersen,36.7,19.3,193.0,3450.0,FEMALE


In [ ]:
print(f"Dataset Shape: {df.shape[0]} rows, {df.shape[1]} columns\n")

df.info()

Dataset Shape: 344 rows, 7 columns

<class 'pandas.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   species            344 non-null    str    
 1   island             344 non-null    str    
 2   culmen_length_mm   342 non-null    float64
 3   culmen_depth_mm    342 non-null    float64
 4   flipper_length_mm  342 non-null    float64
 5   body_mass_g        342 non-null    float64
 6   sex                334 non-null    str    
dtypes: float64(4), str(3)
memory usage: 18.9 KB


## <span style="color: #309c42ff">1.3 Feature Matrix (X) & Target Vector (y) Separation</span>

<blockquote style="border-left: 3px solid #FD1D1D; padding-left: 12px; margin-left: 0;">

In Scikit-Learn conventions:
- <b>Feature Matrix ($X$):</b> A 2D DataFrame containing predictor features.
- <b>Target Vector ($y$):</b> A 1D Series containing ground-truth targets (here, <code>species</code>).

</blockquote>

In [ ]:
target_col = "species"

X = df.drop(columns=[target_col])
y = df[target_col]

print(f"Feature Matrix (X) Shape: {X.shape}")
print(f"Target Vector (y) Shape: {y.shape}")

Feature Matrix (X) Shape: (344, 6)
Target Vector (y) Shape: (344,)


## <span style="color: #309c42ff">1.4 Train-Test Splitting (80/20 Ratio)</span>

<blockquote style="border-left: 3px solid #FD1D1D; padding-left: 12px; margin-left: 0;">

We partition the dataset into 80% training data and 20% test data.
- <code>test_size=0.2</code>: Allocates 20% of samples for evaluation.
- <code>random_state=42</code>: Fixes the random seed to guarantee identical splits across executions.

</blockquote>

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

print("=== Data Partition Summary ===")
print(f"Total Samples  : {len(df)}")
print(f"X_train Shape  : {X_train.shape} ({len(X_train)/len(df):.0%} of dataset)")
print(f"X_test Shape   : {X_test.shape} ({len(X_test)/len(df):.0%} of dataset)")
print(f"y_train Shape  : {y_train.shape}")
print(f"y_test Shape   : {y_test.shape}")

=== Data Partition Summary ===
Total Samples  : 344
X_train Shape  : (275, 6) (80% of dataset)
X_test Shape   : (69, 6) (20% of dataset)
y_train Shape  : (275,)
y_test Shape   : (69,)


## <span style="color: #309c42ff">1.5 Why must a model never see the test set during training?</span>

<blockquote style="border-left: 3px solid #FD1D1D; padding-left: 12px; margin-left: 0;">

Because the model would memorize the patterns in the data and won't be predicting anymore, the test data would be useless and can be considered just training data at this point. The model needs data that it didn't see before to test its performance.

</blockquote>


## <span style="color: #309c42ff">1.6 Summary & Key Takeaways</span>

<blockquote style="border-left: 3px solid #2e1a9aff; padding-left: 12px; margin-left: 0;">

📌 <b>Summary & Next Steps:</b>
- ✅ <b>Data Ingestion:</b> Palmer Penguins dataset loaded successfully (344 rows, 7 columns).
- ✅ <b>Target Definition:</b> Selected <code>species</code> as target $y$ and remaining attributes as features $X$.
- ✅ <b>Split Execution:</b> Partitioned data into 275 training samples and 69 test samples.
- 🚀 <b>Next (Day 2):</b> Preprocessing numerical features (scaling) and handling categorical attributes (one-hot encoding) prior to model training.

</blockquote>